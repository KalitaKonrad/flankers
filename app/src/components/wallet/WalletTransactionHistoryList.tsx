import React from 'react';
import { FlatList, ListRenderItemInfo, StyleSheet } from 'react-native';
import { List } from 'react-native-paper';

import { shuffle } from '../../utils/shuffle';
import { AppText } from '../shared/AppText';

enum TransactionType {
  MATCH_WIN,
  MATCH_LOSS,
  TOP_UP,
}

interface TransactionRecord {
  amount: number;
  type: TransactionType;
}

const TRANSACTION_TYPE_TO_AMOUNT_SIGN = {
  [TransactionType.TOP_UP]: '+',
  [TransactionType.MATCH_WIN]: '+',
  [TransactionType.MATCH_LOSS]: '-',
};

const TRANSACTION_TYPE_TO_NAME = {
  [TransactionType.TOP_UP]: 'DoÅ‚adowanie',
  [TransactionType.MATCH_WIN]: 'Wygrany mecz',
  [TransactionType.MATCH_LOSS]: 'Przegrany mecz',
};

const TRANSACTION_TYPE_TO_ICON = {
  [TransactionType.TOP_UP]: 'bank-plus',
  [TransactionType.MATCH_WIN]: 'trophy',
  [TransactionType.MATCH_LOSS]: 'trophy-broken',
};

const mockLosses = Array(5).fill({
  amount: 5,
  type: TransactionType.MATCH_LOSS,
});
const mockWins = Array(5).fill({ amount: 10, type: TransactionType.MATCH_WIN });
const mockTopUps = Array(5).fill({ amount: 15, type: TransactionType.TOP_UP });

const mockData: TransactionRecord[] = shuffle([
  ...mockLosses,
  ...mockWins,
  ...mockTopUps,
]);

export const WalletTransactionHistoryList: React.FC = () => {
  const renderItem = ({ item }: ListRenderItemInfo<TransactionRecord>) => (
    <List.Item
      title={`${
        TRANSACTION_TYPE_TO_AMOUNT_SIGN[item.type]
      }${item.amount.toFixed(2)}PLN`}
      right={() => <AppText>{TRANSACTION_TYPE_TO_NAME[item.type]}</AppText>}
      left={(props) => (
        <List.Icon {...props} icon={TRANSACTION_TYPE_TO_ICON[item.type]} />
      )}
    />
  );

  return (
    <FlatList
      contentContainerStyle={styles.contentContainer}
      data={mockData}
      renderItem={renderItem}
    />
  );
};

const styles = StyleSheet.create({
  contentContainer: {
    paddingRight: 16,
  },
});
