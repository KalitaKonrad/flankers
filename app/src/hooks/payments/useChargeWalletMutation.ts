import { useMutation, useQueryCache } from 'react-query';

import { QUERY_USER_WALLET } from '../../const/query.const';
import { Wallet } from '../../types/payments/wallet';
import { useAxios } from '../useAxios';

interface ChargeWalletResponse {
  data: Wallet;
}

export const useChargeWalletMutation = () => {
  const axios = useAxios();
  const queryCache = useQueryCache();

  return useMutation(
    async (amount: number) => {
      const response = await axios.post<ChargeWalletResponse>(
        '/wallet/charge',
        {
          amount,
        }
      );
      return response.data.data;
    },
    {
      onSuccess: (wallet) => {
        queryCache.setQueryData([QUERY_USER_WALLET], wallet);
      },
    }
  );
};
