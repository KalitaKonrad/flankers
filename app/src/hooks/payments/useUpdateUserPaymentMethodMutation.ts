import { queryCache, useMutation, useQueryCache } from 'react-query';

import {
  QUERY_USER_HAS_ACTIVE_PAYMENT_METHOD,
  QUERY_USER_WALLET,
} from '../../const/query.const';
import { useAxios } from '../useAxios';

export const useUpdateUserPaymentMethodMutation = () => {
  const axios = useAxios();
  const queryCache = useQueryCache();

  return useMutation(
    (paymentMethodKey: string) =>
      axios.post('payments/setup', {
        payment_method: paymentMethodKey,
      }),
    {
      onSuccess: () => {
        queryCache.setQueryData([QUERY_USER_HAS_ACTIVE_PAYMENT_METHOD], true);
      },
    }
  );
};
