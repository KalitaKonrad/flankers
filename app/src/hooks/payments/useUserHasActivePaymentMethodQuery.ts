import { useQuery } from 'react-query';

import { QUERY_USER_HAS_ACTIVE_PAYMENT_METHOD } from '../../const/query.const';
import { useAxios } from '../useAxios';

interface UserHasActivePaymentMethodResponse {
  data: boolean;
  message: string;
}

export const useUserHasActivePaymentMethodQuery = () => {
  const axios = useAxios();
  return useQuery<boolean>(QUERY_USER_HAS_ACTIVE_PAYMENT_METHOD, async () => {
    const response = await axios.get<UserHasActivePaymentMethodResponse>(
      '/payments/active'
    );
    return response.data.data;
  });
};
