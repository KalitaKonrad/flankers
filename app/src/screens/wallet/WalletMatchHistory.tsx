import React from 'react';
import {
  FlatList,
  SafeAreaView,
  StatusBar,
  StyleSheet,
  TouchableOpacity,
} from 'react-native';
import { List } from 'react-native-paper';

import { Match } from '../../types/match';

interface MatchHistoryProps {
  name: string;
  matchHistory: Match[];
}

interface WalletMatchItemProps {
  item: Match;
  onPress: () => void;
}

const WalletMatchItem: React.FC<WalletMatchItemProps> = ({ item, onPress }) => (
  <TouchableOpacity onPress={onPress} />
);

enum MatchVerdict {
  WIN = 'win',
  LOSS = 'loss',
}

interface MatchOutcome {
  verdict: MatchVerdict;
  moneyAmount: number;
}

const getMatchOutcome = (): MatchOutcome => {
  return { verdict: MatchVerdict.WIN, moneyAmount: 10 };
};

export const WalletMatchHistory: React.FC<MatchHistoryProps> = ({
  name,
  matchHistory,
}) => {
  const matchOutcome = getMatchOutcome();

  return (
    <List.Item
      // TODO: render conditionally match outcome (win/loss etc)
      title="hehe"
      description="Result"
      left={(props) => <List.Icon {...props} icon="star" />}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: StatusBar.currentHeight || 0,
  },
  item: {
    padding: 20,
    marginVertical: 8,
    marginHorizontal: 16,
  },
  title: {
    fontSize: 32,
  },
});
