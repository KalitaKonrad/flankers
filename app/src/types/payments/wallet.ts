export interface Wallet {
  balance: number;
  created_at: Date;
  id: number;
  owner_id: string;
  updated_at: Date;
}
