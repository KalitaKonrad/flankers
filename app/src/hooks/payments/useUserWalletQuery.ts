import { useQuery } from 'react-query';

import { QUERY_USER_WALLET } from '../../const/query.const';
import { Wallet } from '../../types/payments/wallet';
import { useAxios } from '../useAxios';

interface UserWalletResponse {
  data: Wallet;
}

export const useUserWalletQuery = () => {
  const axios = useAxios();
  return useQuery<Wallet>(QUERY_USER_WALLET, async () => {
    const response = await axios.get<UserWalletResponse>('/wallet');
    return response.data.data;
  });
};
