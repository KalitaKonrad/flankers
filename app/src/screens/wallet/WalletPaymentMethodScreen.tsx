import { StackScreenProps } from '@react-navigation/stack';
import React, { useState } from 'react';
import { StyleSheet, View } from 'react-native';
import { CreditCardInput } from 'react-native-credit-card-input';

import { Container } from '../../components/layout/Container';
import { PaddedInputScrollView } from '../../components/layout/PaddedInputScrollView';
import { AppButton } from '../../components/shared/AppButton';
import { useUpdateUserPaymentMethodMutation } from '../../hooks/payments/useUpdateUserPaymentMethodMutation';
import { Stripe } from '../../lib/stripe';
import { CreditCardInputResult } from '../../types/credit-card-input-result';
import { WalletScreenStackParamList } from './WalletScreenStack';
// @ts-ignore

type WalletPaymentMethodScreenProps = StackScreenProps<
  WalletScreenStackParamList,
  'PaymentMethod'
>;

const preparePaymentMethodPayload = (
  card: CreditCardInputResult
): CreatePaymentMethodPayload => {
  const { number, expiry, cvc } = card.values;
  const [exp_month, exp_year] = expiry.split('/');

  return {
    type: 'card',
    card: {
      number,
      cvc,
      exp_month,
      exp_year,
    },
  };
};

export const WalletPaymentMethodScreen: React.FC<WalletPaymentMethodScreenProps> = ({
  navigation,
}) => {
  const [card, setCard] = useState<CreditCardInputResult | null>(null);
  const [updatePaymentMethod] = useUpdateUserPaymentMethodMutation();
  const [isPending, setPending] = useState(false);

  const onSave = async () => {
    if (!card || !card.valid) {
      alert('Wprowadź prawidłowe dane');
      return;
    }

    setPending(true);

    try {
      const response = await Stripe.createPaymentMethod(
        preparePaymentMethodPayload(card)
      );
      const { id } = response.data;
      await updatePaymentMethod(id, { throwOnError: true });
      navigation.navigate('Wallet');
    } catch (error) {
      alert(
        'Podczas zapisywania metody płatności wystąpił błąd. Spróbuj ponownie lub skontaktuj się z supportem.'
      );
    } finally {
      setPending(false);
    }
  };

  return (
    <Container>
      <PaddedInputScrollView>
        <View style={styles.cardContainer}>
          <CreditCardInput
            onChange={(data: CreditCardInputResult) => setCard(data)}
          />
        </View>
        <AppButton
          loading={isPending}
          disabled={isPending}
          mode="contained"
          onPress={onSave}>
          Zapisz metodę płatności
        </AppButton>
      </PaddedInputScrollView>
    </Container>
  );
};

const styles = StyleSheet.create({
  cardContainer: {
    paddingVertical: 64,
  },
});
