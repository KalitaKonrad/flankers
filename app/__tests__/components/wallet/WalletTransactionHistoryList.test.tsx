import { render } from '@testing-library/react-native';
import * as React from 'react';

import { WalletTransactionHistoryList } from '../../../src/components/wallet/WalletTransactionHistoryList';
import { WalletChargeSource } from '../../../src/types/payments/wallet';

describe('<WalletTransactionHistoryList>', () => {
  it('renders correctly when empty data provided', () => {
    const tree = render(
      <WalletTransactionHistoryList transactions={[]} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
