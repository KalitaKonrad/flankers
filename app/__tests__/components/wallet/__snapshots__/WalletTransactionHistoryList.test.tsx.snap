// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`< WalletTransactionHistoryList> renders correctly when all data is provided 1`] = `
<RCTScrollView
  ListEmptyComponent={
    <withTheme(List.Item)
      accessibilityComponentType={true}
      accessibilityTraits=""
      title="Brak transakcji"
    />
  }
  contentContainerStyle={
    Object {
      "paddingRight": 16,
    }
  }
  data={
    Array [
      Object {
        "amount": "string",
        "created_at": 2020-12-31T23:00:00.000Z,
        "source": "game_won",
        "wallet_id": 5,
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            false,
            Array [
              Object {
                "padding": 8,
              },
              undefined,
            ],
          ]
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "margin": 8,
                  "width": 40,
                },
                Object {
                  "marginLeft": 0,
                  "marginRight": 16,
                  "marginVertical": 0,
                },
              ]
            }
          >
            <Text />
          </View>
          <View
            style={
              Array [
                Object {
                  "marginVertical": 6,
                  "paddingLeft": 8,
                },
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                    },
                    undefined,
                  ],
                ]
              }
            >
              +NaNPLN
            </Text>
          </View>
          <View
            style={
              Object {
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "fontSize": 14,
                    "fontWeight": "normal",
                  },
                  undefined,
                ]
              }
            >
              Wygrany mecz
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`< WalletTransactionHistoryList> renders correctly when empty data provided 1`] = `
<RCTScrollView
  ListEmptyComponent={
    <withTheme(List.Item)
      accessibilityComponentType={true}
      accessibilityTraits=""
      title="Brak transakcji"
    />
  }
  contentContainerStyle={
    Object {
      "paddingRight": 16,
    }
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onLayout={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          false,
          Array [
            Object {
              "padding": 8,
            },
            undefined,
          ],
        ]
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "marginVertical": 6,
                "paddingLeft": 8,
              },
              Object {
                "flex": 1,
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "fontSize": 16,
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                  },
                  undefined,
                ],
              ]
            }
          >
            Brak transakcji
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
